# Notes

For each species, manual curation is necessary to produce a usable sample sheet (in most cases). These are processing notes on each species.

At the moment, there is no automated way to add publication data for each metadata file, and so this is manually tracked via Google Sheet.

### Workflow

To add a new species, you'll need to do the following:

First, run the setup blocks. Input and output of sra metadata should be done using the get_sra and write_metadata functions, which standardize formats. Note that the paths are relative, and by default R markdown will use the location of the document as the working directory. So you shouldn't need to change any paths as long as you don't move files around.

The get_sra function takes a BioProject ID (or species, but better to use BioProjects in most cases), and reads in the default columns from the sample metadata file (BioProjet, BioSample, Run, Library Name, Organism, sex). You'll likely want to read in additional columns, but which changes from dataset to dataset. Generally I try to capture location / population information, but for the moment have not been worrying about much else, except fields I'll need to filter on later (to remove domesticated samples, pooled samples, etc). If locality information is available, fill in country and continent where possible. I also keep the original submitter ID, sample name, alias, whatever it is called, but rename to Submitter_Id for consistency.

It helps to look at the metadata file in Excel or a text editor to see what may be useful to read in, and also check the original paper to see what metadata they provide. Once you know what you want, specify the additional columns as a character vector to the opt_cols argument.

Now, you'll have the metadata from SRA. In some cases, you are done. In other cases you'll need to parse supplemental data from papers. The readxl (Excel spreadsheets), tabulizer (PDF tables), and rvest (html) packages are helpful here; the goal is that you should not need to manually touch anything. I try to use PMC links rather than journal links as they are probably more stable. Occasionally I will hard code information from figures as needed to fill in e.g. populations or similar information.

Then you'll need to merge the sra and paper metadata, which can be tricky. If you really can't figure it out, and the SRA metadata is not sufficient on its own (the paper indicates multiple populations were sampled but you cannot link those populations to BioSamples), indicate metadata is inadequate and move on. 

Finally, you'll need to do some last cleanup and export. The Organism field is key here, as the export code splits on this field to produce one file per species. Check to make sure this field is correct (some datasets have a separate "Species" field or something that is more accuracte than Organism). Anything only identified to the genus level (e.g., Genus cf. species; Genus spp; Genus) should be renamed to Genus_spp, and all bionomials should be renamed to replace spaces with underscores (since these will become filenames). Then, run the export code, write_metadata. This splits the input dataframe by Organism, and adds a reference genome column based on the reference species indicated in the sp argument and the genome_key file. This function then writes out a run metadata (with Run, LibraryName and other columns for mapping), and a sample metadata (population information and the like). 

*Critical: make sure that if you use select statements in your code, you only remove columns, or if you must use select to pick columns to keep, be sure to keep at least Organism, BioProject, BioSample, Run, and LibraryName, since those are needed by the write_metadata function to produce the sample sheet for SRA download and mapping.*

All done! Don't forget to git commit and git push, with a helpful commit message, such as "Added metadata for species xxxx".

#### New BioProjects

Sometimes, there may be a BioProject that was missed by the SRA parsing code, which is not perfect. You can use RunSelector to manually download needed metadata, and then put file (named appropriately) in the metadata directory, and then parse as normal. Be sure to add BioProject and publication link to spreadsheet. Note that it downloads as a csv and the read function expects tsv, so you'll need to clean up. There is a script, cleanup_runselector.R, that can help.

### Setup

First we load libraries

```{r}
library(rvest)
library(stringr)
library(tabulizer)
library(readxl)
library(tidyverse)
library(docxtractr)

```

Then we set some default processing. For all datasets, we need the BioProject (to link to publication), the BioSample (to link to VCF), and Organism. 

Paths for input and output:

```{r}
clean_metadata_path <- "cleaned-metadata/"
raw_metadata_path <- "SRA-sample-metadata/"
```

A genome key file that links reference species to GenBank accession:

```{r}
genome_key_file <- "Organism_Metadata.tsv"
genome_key <- read_tsv(file=genome_key_file)
```

Functions for reading and writing metadata:

```{r}

get_sra <- function(sp, path=raw_metadata_path, opt_cols) {
  default_cols = c("BioProject", "BioSample", "Run", "Library Name", "Organism", "sex")
  if (missing(opt_cols)) {
    cols = enquo(default_cols)
  } else {
    default_cols = c(default_cols, opt_cols)
    cols = enquo(default_cols)
  }
  read_tsv(str_c(raw_metadata_path, "SRA_Metadata_", sp, ".tsv")) %>% 
    select(!!cols) %>% rename(LibraryName = `Library Name`)
}

write_metadata <- function(df, refSp) {
  refGenome <- genome_key %>% filter(Organism == refSp) %>% pull(AssemblyAccession)
  
  #Run metadata
  
  df %>% mutate(refGenome = refGenome) %>% 
    select(Organism, BioProject, BioSample, Run, LibraryName, refGenome) %>% distinct() %>%
    split(., .$Organism) %>%
    imap(~ write_csv(as.data.frame(.x), file = str_c(clean_metadata_path, .y, '_run_metadata.csv')))
  
  #Sample metadata
  
  df %>% mutate(refGenome = refGenome) %>% 
    select(-Run, -LibraryName) %>% distinct() %>%
    split(., .$Organism) %>%
    imap(~ write_csv(as.data.frame(.x), file = str_c(clean_metadata_path, .y, '_sample_metadata.csv')))

}
  
```

## METADATA CODE BELOW ##

### Amphilophus citrinellus

Sample metadata downloaded from: https://datadryad.org/stash/dataset/doi:10.5061/dryad.bcc2fqz91

Keyed on sample it / submitter id.


```{r}
sp<-"Amphilophus_citrinellus"
bp<-"PRJEB38173"

#metadata from paper
dryad_link<-"https://datadryad.org/stash/downloads/file_stream/561311"
paper<-read_csv(dryad_link) %>% select(Sample_name, Species, Lake, Sampling_Year)

sra<-get_sra(bp, opt_cols = c("Submitter_Id")) %>%
  mutate(Submitter_Id = str_remove(Submitter_Id, "_WGS")) %>% 
  inner_join(paper,by=c("Submitter_Id" = "Sample_name")) %>%
  mutate(Organism = case_when(
    Species == "Amphilophus cf. citrinellus" ~ "Amphilophus spp",
    Species == "Amphilophus spp. (hybrid)" ~ "Amphilophus spp",
    TRUE ~ Species
   )) %>%
  select(-Species) %>%
  mutate(Organism = str_replace_all(Organism, " ", "_"))

write_metadata(sra, sp)
```

### Anas platyrhynchos

Sample metadata embedded in PDF in supplemental material: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6050300/bin/41467_2018_4868_MOESM1_ESM.pdf

Location of mallard sampling taken from paper text.

```{r}
sp<-"Anas_platyrhynchos"
bp<-"PRJNA450892"

sra<-get_sra(bp, opt_cols = c("Sample_Type", "sex")) %>% 
  mutate(Sample_Type = str_remove(Sample_Type, "_F"))

supmat<-"https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6050300/bin/41467_2018_4868_MOESM1_ESM.pdf"
paper<-do.call(rbind,extract_tables(supmat, pages=c(38:40)))
colnames(paper) <- c("sample", "breed", "mapping_rate", "insert_size", "depth")

meta<-paper %>% as_tibble %>%
  filter(breed != "Breed", breed != "") %>% 
  left_join(sra, by=c("sample" = "Sample_Type")) %>%
  filter(breed == "Mallard") %>%
  mutate(Continent = "Asia", Country = "China", Locality = "Aoji Duck Farm, Zhejiang Province") %>%
  select(-mapping_rate, -insert_size, -depth)

write_metadata(meta, sp)

```

### Anguilla anguilla

Metadata provided to SRA is great and totally complete, no need to dig into paper.
Generally this is very low coverage data, with only a fraction of individuals above 5x, so may not be great.
Closely related species Anguilla rostrata also available

```{r}
sp <- "Anguilla_anguilla"
bp <- "PRJNA668259"

sra<-get_sra(bp, opt_cols = c("Isolate", "Ecotype", "Collection_date", "lat_lon", 
                              "geo_loc_name_country", "geo_loc_name_country_continent", 
                              "geo_loc_name")) %>%
  rename(Country=geo_loc_name_country, 
         Continent=geo_loc_name_country_continent,
         Locality=geo_loc_name)

write_metadata(sra, sp)
```

### Anolis carolinensis

Single individuals of A. porcatus and A. allisoni in same BioProject.

Metadata is okay, but need to get more info about location from the paper: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6681179/

Locked in a Zip archive so need to process

```{r}
sp <- "Anolis_carolinensis"
bp <- "PRJNA533001"
  
sra<-get_sra(bp, opt_cols = c("Sample Name", "Sample_Type")) %>% 
  rename(Submitter_Id = `Sample Name`, Isolate = Sample_Type)

supmat<-"https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6681179/bin/evz110_supplementary_data.zip"
temp <- tempfile()
download.file(supmat,temp)
unzip(temp, files=c("Table_S1.xlsx"), junkpaths=TRUE, exdir=raw_metadata_path)
paper<-read_excel(str_c(raw_metadata_path, "Table_S1.xlsx")) %>% select(Sample:Longitude)
unlink(temp)

meta<-inner_join(paper, sra, by=c("Sample" = "Submitter_Id"))

write_metadata(meta, sp)

```

### Astatotilapia calliptera

This dataset is complex, spanning many species in the cichild flock. Best to process by BioProject, not species. The metadata is also terrible and poorly organized / hard to parse.

Will need to think a bit about whether this is usable, but the metadata is here if needed.

```{r}
sp="Astatotilapia_calliptera"
bp="PRJEB1254"

split_biosample <- function(text) {
   seq2 <- Vectorize(seq.default, vectorize.args = c("from", "to"))
   split_input <- as_tibble(str_split(text, "-", simplify = TRUE)) %>% 
     mutate(V2 = ifelse(V2 == "", V1, V2)) %>%
     mutate(output = sapply(seq2(from=str_remove(V1, "SAMEA"), to=str_remove(V2, "SAMEA"), by=1), paste, sep="", collapse=",")) %>%
     pull(output)
 }


sra<-get_sra(bp, opt_cols = c("Alias")) %>% 
  rename(Submitter_Id = Alias)

supmat1<-"https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4700518/bin/NIHMS66047-supplement-01.pdf"
table16<-do.call(rbind,extract_tables(supmat1, pages=c(47:47)))
colnames(table16) <- c("SampleName", "BioSample")

paper1<-as_tibble(table16) %>% separate_rows(BioSample, sep=",") %>% filter(SampleName != "Samples") %>%
  filter(BioSample != "") %>%
  separate_rows(BioSample, sep=",") %>%
  mutate(BioSample = str_remove(BioSample, "\\r")) %>%
  mutate(BioSample = trimws(BioSample)) %>%
  mutate(BioSample = split_biosample(BioSample)) %>%
  separate_rows(BioSample) %>% select(BioSample, SampleName) %>%
  mutate(BioSample = str_c("SAMEA", BioSample))

supmat2<-"https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6443041/bin/NIHMS82209-supplement-Supplementary_information.pdf"
table4 <- do.call(rbind,extract_tables(supmat2, pages=c(50:50)))
colnames(table4) <- c("SampleName", "BioSample")

paper2<-as_tibble(table4) %>% filter(SampleName != "Samples") %>%
  mutate(SampleNameClean = c(rep("Lake Malawi populations",11), rep("Lake Malawi trios",3), 
                        rep("Malawi catchment", 2), rep("Indian ocean catchments", 2),
                        rep("Outgroup", 6))) %>%
  filter(BioSample != "") %>%
  separate_rows(BioSample, sep=",") %>%
  mutate(BioSample = str_remove(BioSample, "A. stuartgranti")) %>%
  mutate(BioSample = str_remove(BioSample, "L. lethrinus")) %>%
  mutate(BioSample = str_remove(BioSample, "A. calliptera Salima")) %>%
  mutate(BioSample = trimws(BioSample)) %>%
  filter(str_detect(BioSample, "SAMEA")) %>%
  mutate(BioSample = str_replace(BioSample, "SAMEA1904329-", "SAMEA1904329-SAMEA1904331")) %>%
  mutate(BioSample = str_replace(BioSample, "SAMEA1904350-", "SAMEA1904350-SAMEA1904353")) %>%
  mutate(BioSample = trimws(BioSample)) %>%
  mutate(BioSample = split_biosample(BioSample)) %>%
  separate_rows(BioSample) %>% select(BioSample, SampleNameClean) %>%
  mutate(BioSample = str_c("SAMEA", BioSample))

meta<-full_join(paper1, paper2, by=c("BioSample" = "BioSample")) %>% distinct() %>% 
  full_join(sra, by=c("BioSample" = "BioSample")) %>% distinct() %>%
  rename(SampleName1 = SampleName, SampleName2 = SampleNameClean) %>%
  filter(!is.na(BioProject))

meta %>% filter(Organism == "Astatotilapia_calliptera" | Organism == "Astatotilapia") %>%
  mutate(Organism = ifelse(Organism == "Astatotilapia", "Astatotilapia_spp", Organism)) %>%
  write_metadata(sp)
```


### Astyanax mexicanus

Beautiful metadata, everything in SRA submission

```{r}
sp="Astyanax_mexicanus"
bp<-"PRJNA260715"
  
sra<-get_sra(bp, opt_cols = c("Sample Name")) %>% 
  rename(Submitter_Id = `Sample Name`) %>%
  separate(Submitter_Id, into=c("Population"), remove=FALSE, extra="drop")

write_metadata(sra, sp)

```

### Athene cunicularia

SRA submission mostly looks good, although there does not appear to be a way to link the SRA populations (A, B, C) to the real world places they are from.

```{r}

sp="Athene_cunicularia"
bp="PRJNA431202"
  
sra<-get_sra(bp, opt_cols = c("Sample Name", "Ecotype")) %>% 
  rename(Submitter_Id = `Sample Name`) %>%
  separate(Ecotype, into=c("Site", "Site_Type", "Year"), remove=FALSE, extra="drop")

write_metadata(sra, sp)
```

### Chaenogobius annularis

Adding information about population defined in the paper (Pacific Ocean vs Sea of Japan). 

Japan:Misaki -> PO
Japan:Misaki -> PO
Japan:Onagawa -> PO
Japan:Onagawa -> PO
Japan:Aomori -> SJ
Japan:Aomori -> SJ
Japan:Aomori -> SJ
Japan:Murakami -> SJ
Japan:Murakami -> SJ
Japan:Murakami -> SJ
Japan:Misaki -> PO
Japan:Onagawa -> PO

(from Table S4 from the paper)

```{r}

sp="Chaenogobius_annularis"
bp="PRJDB7125"
  
sra<-get_sra(bp, opt_cols = c("Isolate", "Collection_date", "geo_loc_name_country", "geo_loc_name_country_continent", "geo_loc_name", "Sample_Name")) %>% 
  rename(Submitter_Id = Sample_Name, Country = geo_loc_name_country, Continent = geo_loc_name_country_continent, Locality = geo_loc_name) %>%
  mutate(Population = case_when(
    Locality == "Japan:Misaki" | Locality == "Japan:Onagawa" ~ "Pacific Ocean",
    Locality == "Japan:Aomori" | Locality == "Japan:Murakami" ~ "Sea of Japan"    
  ))

write_metadata(sra, sp)

```


### Chloebia gouldiae

Very nice SRA metadata

```{r}

sp="Chloebia_gouldiae"
bp="PRJNA515277"
  
sra<-get_sra(bp, opt_cols = c("Ecotype", "Collection_date", "geo_loc_name_country", "geo_loc_name_country_continent", "geo_loc_name", "lat_lon", "Strain", "Cultivar")) %>% 
  rename(Submitter_Id = Cultivar, Country = geo_loc_name_country, Continent = geo_loc_name_country_continent, Locality = geo_loc_name) %>%
  select(-sex) %>%
  rename(sex = Strain)

write_metadata(sra, sp)

```

### Clupea spp.

Complicated by the fact that different metadata fields are used for different BioProjects; will probably need to parse by BioProject and then combined to avoid lots of complicated case_when statements.

Parse by BioProject, then output by species.

Additional metadata could be added froom "meta1" and "meta2" tables, but it is a confusing mess and I haven't done this yet.


```{r}

#PRJNA642736 filter to keep only Sample_Type == "individual"
sp="Clupea_harengus"

sra1<-get_sra("PRJNA642736", opt_cols = c("Ecotype", "Sample_Type")) %>% 
  rename(Locality = Ecotype) %>%
  mutate(Submitter_Id = LibraryName) %>% 
  filter(Sample_Type == "individual") %>%
  select(-Sample_Type) %>% select(Organism, BioProject, Run, LibraryName, BioSample, sex, Submitter_Id, Locality)

temp <- tempfile()
supmat<- "https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7738190/bin/elife-61076-supp2.xlsx"
download.file(supmat,temp)
meta1<-read_excel(temp, skip=1, col_types = "text", trim_ws = TRUE) 
unlink(temp)

#PRJNA338612 filter to keep only Sample_Type == "individual"

#grab pop info from paper

paper2<-"https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5410801/pdf/pnas.201617728.pdf"
table1 <- extract_tables(paper2, pages=c(3:3))[[2]]
colnames(table1) <- c("Locality", "N", "blank", "lat", "lon", "salinity", "date", "spawn_season")
meta2<-as_tibble(table1) %>% 
  filter(date != "") %>% 
  select(-blank, -N) %>% 
  separate(Locality, into=c("Locality", "Population"), remove=FALSE, sep="[†*]") %>% 
  mutate(Population = trimws(str_replace(Population, "[Ww]", ""))) %>%
  mutate(Population = str_replace(Population, "[Ä]", "A"))

sra2<-get_sra("PRJNA338612", opt_cols = c("Sample Name", "Sample_Type")) %>% 
  rename(Submitter_Id = `Sample Name`) %>%
  filter(Sample_Type == "Individual DNA") %>%
  select(-Sample_Type) %>% separate(Submitter_Id, into=c("Population", "Sample"), sep=-1, remove=FALSE) %>%
  left_join(meta2, by=c("Population" = "Population")) %>% select(Organism, BioProject, Run, LibraryName, BioSample, sex, Submitter_Id, Locality)

sra3<-get_sra("PRJNA356817", opt_cols = c("Sample Name")) %>% 
  rename(Submitter_Id = `Sample Name`) %>%
  mutate(Locality = case_when(
    Submitter_Id == "AAPF" | Submitter_Id == "AAPM" ~ "Norway:Bergen",
    Submitter_Id == "ABPF" | Submitter_Id ==  "ABPM" ~ "Sweden:Hästskär",
    TRUE ~ NA_character_)) %>%
    filter(!is.na(Locality))  %>% select(Organism, BioProject,  Run, LibraryName, BioSample, sex, Submitter_Id, Locality)

#export
bind_rows(sra1, sra2, sra3) %>%
  write_metadata(sp)

```

### Coilia nasus

Sample metadata is in a word document, will need to find a way to parse that...
Paper: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6939831/

```{r}

sp="Coilia_nasus"
bp="PRJNA422339"

sra<-get_sra(bp) %>%
  mutate(Submitter_Id = LibraryName)

supmat<-"https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6939831/bin/giz157_supplemental_figures_and_tables.zip"
temp <- tempfile()
download.file(supmat,temp)
unzip(temp, files=c("Supplementary materials revised.docx"), junkpaths=TRUE, exdir=raw_metadata_path)
table8<-read_docx(str_c(raw_metadata_path, "Supplementary materials revised.docx")) %>% 
  docx_extract_tbl(8) %>% 
  select(Library, Sample) %>%
  separate(Sample, into=c("Population", "Sample"))
unlink(temp)

#need to remove trailing numbers from population and drop Sample
table8 <- mutate(table8, Population = str_remove(Population, "[0-9]+$")) %>% select(-Sample)

table1<-extract_tables("https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6939831/pdf/giz157.pdf", pages=c(4))[[1]] 
colnames(table1)<-table1[1,]
table1[4:6,1] <- "River"
table1[8,1] <- "Lake"
table1<-as_tibble(table1[2:8,]) %>% right_join(table8, by=c("Sample" = "Population")) %>%
  rename(Population = Sample)

meta<-left_join(sra, table1, by=c("Submitter_Id" = "Library"))

write_metadata(meta, sp)
```


### Corvus species

BioProject: PRJNA192205
Papers: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5095515/, https://www.ncbi.nlm.nih.gov/pubmed/24948738
Genome: Corvus_brachyrhynchos or Corvus_cornix_cornix

```{r}

sp <- "Corvus_cornix_cornix"
bp <- "PRJNA192205"

sra<-get_sra(bp, opt_cols=c("Sample Name", "geo_loc_name_country", "geo_loc_name_country_continent", "geo_loc_name")) %>%
  rename(Submitter_Id = `Sample Name`, Country = geo_loc_name_country, Continent = geo_loc_name_country_continent, Locality=geo_loc_name)

write_metadata(sra, sp)
```


### Coryphaenoides rupestris

PRJNA430094	
https://pubmed.ncbi.nlm.nih.gov/29507380/


```{r}
sp<-"Coryphaenoides_rupestris"
bp<-"PRJNA430094"

sra<-get_sra(bp, opt_cols = c("Sample Name", "geo_loc_name_country", "geo_loc_name_country_continent", "geo_loc_name", "lat_lon", "Depth")) %>%
  rename(Submitter_Id = `Sample Name`, Country = geo_loc_name_country, Continent = geo_loc_name_country_continent, Locality=geo_loc_name)

write_metadata(sra, sp)

```


### Coturnix japonica	

PRJNA339911	https://academic.oup.com/gigascience/article/7/5/giy049/4995262

```{r}
sp<-"Coturnix_japonica"
bp<-"PRJNA339911"

supmat<-"https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5961004/bin/giy049_supp.zip"
temp <- tempfile()
download.file(supmat,temp)
unzip(temp, files=c("Additional file 1-updated.docx"), junkpaths=TRUE, exdir=raw_metadata_path)
table11<-read_docx(str_c(raw_metadata_path, "Additional file 1-updated.docx")) %>% 
  docx_extract_tbl(11) %>% 
  select(Submitter_Id = Sample, Description, source) %>%
  filter(Description == "Wild quail") %>%
  select(-Description) %>%
  mutate(Submitter_Id = str_remove(Submitter_Id, "-[MF]$")) %>%
  mutate(Submitter_Id = str_replace(Submitter_Id, "-", "_"))
unlink(temp)

#paper metadata and SRA metadata disagree on sex of samples, so change sex to missing

sra<-get_sra(bp, opt_cols=c("Sample Name")) %>%
  mutate(sex = "missing") %>%
  rename(Submitter_Id = `Sample Name`) %>%
  mutate(Submitter_Id = str_remove(Submitter_Id, "_[MF]_new$"))
  
meta<-left_join(table11, sra, by=c("Submitter_Id" = "Submitter_Id"))
write_metadata(meta, sp)

```


### Cyprinus carpio	

PRJNA510861 https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6789147/

SRA metadata seems complete enough, although can't find any mention of the resequencing dataset in the PMC link

```{r}
bp<-"PRJNA510861"
sp<-"Cyprinus_carpio"

sra<-get_sra(bp, opt_cols=c("Sample Name", "Breed", "Strain")) %>%
  rename(Submitter_Id = `Sample Name`)

write_metadata(sra, sp)

```


### Dicentrarchus labrax	

PRJNA472842	https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6023918/
PRJNA628166	https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7293100/

There is more precise location data for these samples, but only reported as a figure with no table linking to samples, so can't use it.


```{r}
sp<-"Dicentrarchus_labrax"
bp1<-"PRJNA472842"
bp2<-"PRJNA628166"

sra1<-get_sra(bp1, opt_cols=c("Sample Name", "Ecotype")) %>% rename(Submitter_Id = `Sample Name`)
sra2<-get_sra(bp2, opt_cols=c("Sample Name", "Ecotype")) %>% rename(Submitter_Id = `Sample Name`)
sra<-bind_rows(sra1, sra2)

sra<-mutate(sra, Subpopulation=case_when(
  str_detect(Submitter_Id, "^AT") ~ "Atlantic",
  str_detect(Submitter_Id, "^WM") ~ "W_Mediterranean",
  str_detect(Submitter_Id, "^SE") ~ "SE_Mediterranean",
  str_detect(Submitter_Id, "^NE") ~ "NE_Mediterranean",
  TRUE ~ "Unknown")
) %>%
  rename(Population = Ecotype)

write_metadata(sra, sp)

```

### Egretta garzetta	

PRJNA309350	https://genomebiology.biomedcentral.com/articles/10.1186/s13059-014-0557-1

```{r}
sp <- "Egretta_garzetta"
bp <- "PRJNA309350"

sra <- get_sra(bp, opt_cols = c("geo_loc_name")) %>%
  separate(geo_loc_name,into=c("Country","Locality")) %>%
  mutate(Continent = "Asia")
  
write_metadata(sra, sp)
```

### Eopsaltria australis	

PRJNA476023   https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6736294/

SRA metadata very good, except info for genome species is absent, grab from Excel

```{r}
bp<-"PRJNA476023"
sp<-"Eopsaltria_australis"


temp <- tempfile()
supmat<- "https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6736294/bin/giz111_supplemental_table1.xlsx"
download.file(supmat,temp)
meta1<-read_excel(temp, skip=1, col_types = "text", trim_ws = TRUE) %>%
  mutate(ID = str_replace(ID, " ", "-"))
unlink(temp)

sra<-get_sra(bp, opt_cols=c("geo_loc_name_country",	"geo_loc_name_country_continent",	
                            "geo_loc_name", "lat_lon", "Sample Name")) %>%
  rename(Submitter_Id = `Sample Name`)

meta <- full_join(sra, meta1, by=c("Submitter_Id" = "ID")) %>%
  select(BioProject, BioSample, Run, LibraryName, Submitter_Id, Organism, sex, Country = geo_loc_name_country, 
         Continent = geo_loc_name_country_continent, Location = geo_loc_name, lat_lon, 
         Latitude, Longitude, Population, State) %>%
  mutate(Country = ifelse(is.na(Country), "Australia", Country),
         Continent = ifelse(is.na(Continent), "Oceania", Continent),
         Location = case_when(
           State == "VIC" & Location != "Australia:Victoria" ~ "Australia:Victoria",
           State == "VIC" & is.na(Location) ~ "Australia:Victoria",
           TRUE ~ Location
         ),
         lat_lon = ifelse(is.na(lat_lon), 
                         str_c(abs(round(as.numeric(Latitude),3)), " S", " ", round(as.numeric(Longitude),3), " E"), 
                         lat_lon)) %>%
  select(-Latitude, -Longitude)

write_metadata(meta, sp)

```


### Falco peregrinus

PRJNA686418
https://www.nature.com/articles/s41586-021-03265-0
	
```{r}
sp<-"Falco_peregrinus"
bp<-"PRJNA686418"

sra<-get_sra(bp, opt_cols = c("Sample Name")) %>%
  rename(Submitter_Id = `Sample Name`) %>%
  mutate(Population = str_remove(Submitter_Id, "[\\-0-9]+$"))

write_metadata(sra, sp)

```

### Fidecula species

PRJEB11502	Ficedula albicollis	https://pubmed.ncbi.nlm.nih.gov/26649993/
PRJEB12616	Ficedula albicollis	https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4878770/
PRJEB2984   Ficedula spp https://www.nature.com/articles/nature11584
PRJEB7359	Ficedula spp	https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4617962/



```{r}
sp<-"Ficedula_albicollis"

##Bioproject 1##

bp<-"PRJEB7359"

sra1<-get_sra(bp, opt_cols=c("geo_loc_name_country", "geo_loc_name_country_continent", "Title", "Description")) %>%
  rename(Submitter_Id = Title, Country = geo_loc_name_country, Continent = geo_loc_name_country_continent) %>%
  separate(Submitter_Id, into=c("Population", "Sample"), remove=FALSE, extra="merge", fill = "right")

#ugly clean up because some things are missing information
#source of NRM996601 https://www.gbif.org/occurrence/2248760844
#source of pop key is Fig1 at https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4617962/ as it doesn't seem to be provided as text anywhere

sra1 <- sra1 %>%
  mutate(Country = case_when(
    Population == "I" ~ "Italy",
    Population == "OC" ~ "Sweden",
    Population == "NRM996601" ~ "Sweden",
    Population == "SP" ~ "Sweden",
    Population == "FHYP" ~ "Unknown",
    TRUE ~ Country
  )) %>%
  mutate(Continent = case_when(
    Country == "Italy" ~ "Europe",
    Country == "Sweden" ~ "Europe",
    Country == "Unknown" ~ "Unknown",
    TRUE ~ Continent
  )) %>% 
  mutate(Locality = case_when(
    Population == "SP" ~ "Uppsala",
    Population == "OC" ~ "Öland",
    Population == "OP" ~ "Öland",
    TRUE ~ "Unknown"
  ))

#last cleanup of NRM996601 and removing description field

sra1 <- sra1 %>%
  mutate(Population = ifelse(Population == "NRM996601", "FPAR", Population)) %>%
  mutate(Sample = ifelse(is.na(Sample), Submitter_Id, Sample)) %>%
  select(-Description)

#for reference later
pop_meta <- sra1 %>% select(Country, Continent, Population, Locality, Organism) %>% distinct()

##Bioproject 2##


bp<-"PRJEB12616"

#keeping only parents based on pedigree in Fig 1 here: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4878770/

sra2<-get_sra(bp, opt_cols=c("Alias")) %>%
  rename(Submitter_Id = Alias) %>%
  mutate(Sample = str_remove(Submitter_Id, "[0-9]*$")) %>%
  mutate(Country = "Sweden", Continent = "Europe", Locality = "Öland") %>%
  filter(Sample == "CJ" | Sample == "BL" | Sample == "BY") %>%
  mutate(Population = "OC")

##Bioproject 3##

bp<-"PRJEB11502"

#keeping only parents based on pedigree in Fig 1 here: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4878770/

sra3<-get_sra(bp, opt_cols=c("Alias")) %>%
  rename(Submitter_Id = Alias) %>%
  mutate(Country = "Sweden", Continent = "Europe", Locality = "Öland", Population = "OC")

#some extraneous stuff in the BioProject for some reason, so filter to keep only samples from paper to be sure the metadata is correct

samples<-read_delim("https://datadryad.org/stash/downloads/file_stream/332334", delim=" ")

sra3 <- filter(sra3, Submitter_Id %in% samples$id) %>% mutate(sex = "male")

## Biorproject 4 ## 

bp<-"PRJEB2984"
sra4<-get_sra(bp, opt_cols=c("LibrarySource", "Title")) %>%
  filter(LibrarySource == "GENOMIC") %>% 
  rename(Submitter_Id = Title) %>% select(-LibrarySource) %>% distinct() %>%
  mutate(Population = case_when(
    Organism=="Ficedula_albicollis" ~ "OC",
    Organism=="Ficedula_hypoleuca" ~ "SP"
  )) %>%
  left_join(pop_meta, by = c("Population" = "Population"))

meta <- bind_rows(sra1, sra2, sra3, sra4)

write_metadata(meta, sp)

```

### Fundulus heteroclitus

PRJNA323589   http://science.sciencemag.org/content/354/6317/1305.full


SRA is clean and nice, could add phenotype from Table S1 though


```{r}
bp<-"PRJNA323589"
sp<-"Fundulus_heteroclitus"

sra<-get_sra(bp, opt_cols=c("Sample Name", "Ecotype", "lat_lon", "Collection_date")) %>%
  rename(Submitter_Id = `Sample Name`) %>%
  mutate(Phenotype = case_when(
    Ecotype == "New Bedford Harbor" | Ecotype == "Bridgeport" | Ecotype == "Newark" | Ecotype == "Elizabeth River" ~ "Tolerant", 
    Ecotype == "Block Island" | Ecotype == "Flax Pond" | Ecotype == "Sandy Hook" | Ecotype == "Kings Creek" ~ "Sensitive", 
    TRUE ~ NA_character_
  )) %>%
  rename(Population = Ecotype)

write_metadata(sra, sp)

```

### Gambusia affinis	

PRJNA386810	https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5982815/


```{r}


```

### Gasterosteus species

PRJEB15181	Gasterosteus aculeatus	https://www.ncbi.nlm.nih.gov/pubmed/27569902
PRJDB2374	Gasterosteus wheatlandi	https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3953013/

**more BioProjects may be added -- sort first before doing any processing!**

```{r}


```


###	Haplochromis burtoni 

PRJNA485198   https://www.biorxiv.org/content/10.1101/2021.01.12.426414v1
PRJNA60363    https://www.ncbi.nlm.nih.gov/pubmed/25186727

```{r}


```

### Hippocampus spp

PRJNA612146	Hippocampus comes	https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7889852/	
PRJNA612146	Hippocampus kuda	https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7889852/	

```{r}

```

### Hypoplectrus puella

PRJEB27858    https://www.nature.com/articles/s41559-019-0814-5

```{r}

```


### Larimichthys crocea	

PRJNA629649
https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7645154/

```{r}


```

### Leuciscus waleckii	

PRJNA298374
https://academic.oup.com/mbe/article/34/1/145/2655148

```{r}


```

###	Maylandia zebra	

PRJEB15289
https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6443041/

```{r}

```

###	Nipponia nippon	

PRJNA308878
https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4290368/

```{r}
sp <- "Nipponia_nippon"
bp <- "PRJNA308878"

sra <- get_sra(bp, opt_cols = c("geo_loc_name","Sample Name")) %>%
  separate(geo_loc_name,into=c("Country","Locality")) %>%
  mutate(Continent = "Asia")

write_metadata(sra,sp)

```


### Oreochromis species	

PRJNA59571    Oreochromis niloticus	https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4353498/
PRJDB1657   Oreochromis niloticus	https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4570987/
PRJDB1657	Oreochromis mossambicus	https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4570987/

```{r}

```

### Oryzias latipes	

PRJEB2759	https://www.ncbi.nlm.nih.gov/pubmed/24408034

```{r}

```

### Parus major


PRJNA274877	https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4737754/
PRJNA304164	https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4585896/

There is a supplemental table for PRJNA274877 on page 18 here: "https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4585896/bin/srep14256-s1.pdf", but tabulizer will not read it correctly, even after specifying the area and column locations. So, I classifyed each locality into the appropriate population grouping.

```{r}
sp <- "Parus_major"
bp1 <- "PRJNA274877"
bp2 <- "PRJNA304164"

#First BP
sra1 <- get_sra(bp1, opt_cols=c("birth_location", "Sample Name")) %>%
  separate(birth_location,into=c("Locality","Province"),extra="drop",sep=",") %>%
  separate(Province,sep=" [0-9]",into=c("Province"),extra="drop") %>%
  mutate(Province = str_sub(Province, 2, -1)) %>%
  mutate(Population = case_when(Province == "Sichuan" ~ "EH",
                                Province == "Yunnan" ~ "EH",
                                Province == "Inner Mongolia" ~ "MON",
                                Province == "Mongolia" ~ "MON",
                                Province == "Russia" ~ "MON",
                                Province == "Hubei" ~ "CE",
                                Province == "Anhui" ~ "CE",
                                Province == "Liaoning" ~ "CE",
                                Province == "Gansu" ~ "CE")) %>%
  mutate(Continent = "Asia",
         Country = case_when(Province == "Sichuan" ~ "China",
                                Province == "Yunnan" ~ "China",
                                Province == "Inner Mongolia" ~ "China",
                                Province == "Mongolia" ~ "Mongolia",
                                Province == "Russia" ~ "Russia",
                                Province == "Hubei" ~ "China",
                                Province == "Anhui" ~ "China",
                                Province == "Liaoning" ~ "China",
                                Province == "Gansu" ~ "China"))

#Second BP
temp <- tempfile()
supmat<- "https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4737754/bin/ncomms10474-s3.xlsx"
download.file(supmat,temp)
meta2<-read_excel(temp, skip=1, col_types = "text", trim_ws = TRUE) 
unlink(temp)

sra2 <- get_sra(bp2, opt_cols="Isolate") %>%
  left_join(meta2,by=c("Isolate"="Individual Name")) %>%
  mutate(Country = if_else(Country =="Cambridge, UK","UK",Country)) %>%
  mutate(Country = if_else(Country == "UK", "United Kingdom",Country)) %>%
  select(-Gender,-'Volume(ul)',-'Concentration (ng/ul)') %>%
  mutate(Continent = "Europe") %>%
  rename(Locality=Ciry)

#Combine BPs
sra <- bind_rows(sra1,sra2)

write_metadata(sra,sp)
```

### Passer species

PRJEB27649	Passer domesticus	https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6111181/
PRJNA255814	Passer domesticus	https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5470830/
PRJNA255814	Passer hispaniolensis	https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5470830/
PRJNA255814	Passer italiae	https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5470830/
PRJNA255814	Passer montanus	https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5470830/
*Individual 8887266 is the reference genome, dropping.

PRJEB22939	Passer italiae	https://www.nature.com/articles/s41559-017-0437-7

**The pub was incorrect for PRJNA634228, and I can't find a matching pub to grab locality information for samples. Not including this bioproject for now.
PRJNA634228	Passer montanus	

```{r}
sp <- "Passer_domesticus"
bp1 <- "PRJEB27649"
bp2 <- "PRJNA255814"
bp3 <- "PRJEB22939"
bp4 <- "PRJNA634228"

#BP PRJEB27649
sra1 <- get_sra(bp1, opt_cols=c("Sample_Name","Title")) %>%
  select(-sex)

temp <- tempfile()
supmat <- "https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6111181/bin/rspb20181246supp2.xlsx"
download.file(supmat,temp)
meta1<-read_excel(temp, skip=0, col_types = "text", trim_ws = TRUE) %>%
  select(Individual,Sex,Population,Longitude,Latidue) %>%
  rename(sex=Sex,Latitude=Latidue)
unlink(temp)

sra1 <- left_join(sra1,meta1,by=c("Title"="Individual")) %>%
  separate(Population,into=c("Locality","Country")) %>%
  mutate(Locality = if_else(Locality == "Cape","Cape Verde",Locality)) %>%
  mutate(Country=if_else(Locality == "Tenerife","Canary Islands",Country),
         Country=if_else(Locality == "Cape Verde","Cape Verde",Country),
         Country=if_else(Locality == "Sardinia", "Italy",Country),
         Country=if_else(Locality == "Malta", "Malta",Country)) %>%
  mutate(Continent=case_when(Country == "Canary Islands"~"Africa",
                             Country == "Cape Verde"~"Africa",
                             Country == "Italy"~"Europe",
                             Country == "France"~"Europe",
                             Country == "Kazahkstan"~"Asia",
                             Country == "Norway"~"Europe",
                             Country == "Iran"~"Asia",
                             Country == "Malta"~"Europe")) %>%
  mutate(Sample_Name=Title)

#BP PRJNA255814
sra2 <- get_sra(bp2, opt_cols=c("Library Name","Isolate")) %>%
  filter(Isolate != "8887266") %>%
  separate(LibraryName, into=c("drop1","drop2","drop3","drop4","Sample_ID"),sep="\\.",extra="drop",fill="right") %>%
  select(-drop1,-drop2,-drop3,-drop4) 

suppmat<-"https://advances.sciencemag.org/highwire/filestream/196328/field_highwire_adjunct_files/0/1602996_SM.pdf"
table1 <- extract_tables(suppmat, pages=c(9:10))[[1]]
table2 <- extract_tables(suppmat, pages=c(9:10))[[2]]
colnames(table1) <- colnames(table2) <- c("Species", "Sample_ID", "Sex", "Date", "Location", "Latitude", "Longitude")
table1 <- table1[2:nrow(table1),] %>%
  as_tibble
table2 <- table2 %>%
  as_tibble
table1 <- bind_rows(table1,table2)

#Have to fix typo in table names
table1 <- table1 %>%
  mutate(Sample_ID = if_else(Sample_ID == "Pitaliae_30_Guglionesi_336","Pitaliae_30_Guglionesi_436",Sample_ID))

sra2 <- sra2 %>%
  left_join(table1,by=c("Sample_ID"="Sample_ID")) %>%
  separate(Location, into=c("Locality","Country"),sep=",") %>%
  rename(Sample_Name=Sample_ID) %>%
  mutate(Continent="Europe",
         Locality=str_trim(Locality),
         Country=str_trim(Country)) %>%
  select(-Sex,-Species,-Isolate,-Date)


#BP PRJEB22939
sra3 <- get_sra(bp3,opt_cols=c("geographic_location_(region_and_locality)")) %>%
  rename(Locality=`geographic_location_(region_and_locality)`) %>%
  mutate(Country="Italy",
         Continent="Europe")


sra <- bind_rows(sra1,sra2,sra3) %>%
  mutate(sex=if_else(sex=="M","male",sex),
         sex=if_else(sex=="F","female",sex),
         sex=if_else(sex=="U"|sex=="J","undetermined",sex)) %>%
  rename(Submitter_Id=Sample_Name) %>%
  select(-Title) %>%
  separate(Organism,into=c("genus","species"),extra="drop") %>%
  unite(genus,species,col="Organism")

write_metadata(sra,sp)
```

### Pelecanus crispus

PRJNA212901	Pelecanus crispus	https://www.ncbi.nlm.nih.gov/pubmed/25504712

Only 3 samples, so skipped (intended for use as outgroup)

```{r}

```


### Phylloscopus species

PRJEB21643	Phylloscopus collybita abietinus	https://www.ncbi.nlm.nih.gov/pubmed/29054864
PRJEB21643	Phylloscopus collybita tristis	https://www.ncbi.nlm.nih.gov/pubmed/29054864
PRJNA319295	Phylloscopus trochilus acredula	https://onlinelibrary.wiley.com/doi/abs/10.1002/evl3.15
PRJNA319295	Phylloscopus trochilus trochilus	https://onlinelibrary.wiley.com/doi/abs/10.1002/evl3.15


```{r}


```

###	Pimephales promelas	


PRJNA227290   https://www.ncbi.nlm.nih.gov/pubmed/26513338

```{r}


```


### Plectropomus leopardus	

PRJNA622646	    https://onlinelibrary.wiley.com/doi/abs/10.1111/1755-0998.13207

```{r}


```

### Podarcis muralis

PRJNA515813	    https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6431182/

```{r}

```

### Poecilia species

PRJNA89109	Poecilia formosa	https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5866774/
PRJNA196862	Poecilia latipinna	https://www.nature.com/articles/s41559-018-0473-y
PRJEB10680	Poecilia reticulata	https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7643365/
PRJEB22221	Poecilia reticulata	https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6452659/
PRJEB36450	Poecilia reticulata	https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7643365/
PRJEB39998	Poecilia reticulata	https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7826173/
PRJNA238429	Poecilia reticulata	https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5199103/
PRJNA353986	Poecilia reticulata	https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5290318/
PRJEB10586	Poephila acuticauda	https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4864528/

```{r}


```

### Pungitius pungitius

PRJDB2321	Pungitius pungitius	https://www.ncbi.nlm.nih.gov/pubmed/25818858
PRJNA477377	Pungitius pungitius	https://academic.oup.com/mbe/article/36/1/28/5112968

```{r}

```


### Strix occidentalis caurina

PRJNA338167	https://www.ncbi.nlm.nih.gov/pubmed/28992302

```{r}


```

###Sylvia atricapilla

PRJNA616371	https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7173969/

```{r}


```

###	Symphodus melops

PRJNA354496   https://onlinelibrary.wiley.com/doi/full/10.1111/mec.15310

```{r}


```

### Taeniopygia species
PRJEB10586	Taeniopygia bichenovii	https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4864528/
PRJEB10586	Taeniopygia guttata	https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4864528/

```{r}

```


### Takifugu species


PRJNA522329	Takifugu bimaculatus	https://dx.doi.org/10.21203/rs.2.16620/v1
PRJNA522329	Takifugu rubripes	https://dx.doi.org/10.21203/rs.2.16620/v1

```{r}

```


### Thunnus albacares	

PRJEB14402	https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5408087/

```{r}

```

### Tyto alba

PRJNA212909	Tyto alba	https://www.ncbi.nlm.nih.gov/pubmed/25504712


